README.md
# CDK

Maak een branch voor CDK (Cultuurhuys De Kroon) gerelateerde projecten.

## Introductie

Op dit moment - juni 2022 - zijn er nog geen scripts.

De wijzigingen worden bijgehouden in het document: [Changelog](CHANGELOG.md).

* De structuur wordt opgezet om gebruikt te kunnen worden met [ModuleBuilder](https://github.com/PoshCode/ModuleBuilder) om een module te kunnen maken.
* Het wordt geadviseerd om ModuleBuilder in een van je paden in $env:PSModulePath te zetten. Het liefst in een pad dat je zowel vanuit  PowerShell 'Desktop' en 'Core' kunt benaderen.
* Dus deze repo bevat alleen de bestanden om een module te kunnen *bouwern*. Je dient dus bekend te zijn met _ModuleBuilder_ en de gemaakte module in je eigen omgeving te kunnen gebruiken.

---

## Voorbeeld om een module te bouwen

Pas eerst de volgende 2 parameters in "_FullPathTo_\CultuurhuysDeKroon\Source\Build.psd1" aan.

```PowerShell
OutputDirectory          = "$( $env:OneDrive )\Modules\CultuurhuysDeKroon"
VersionedOutputDirectory = $true
```

* OutputDirectory
  * Die verwijst naar de map waar de gemaakte module geplaatst wordt.
  * Bij voorkeur is deze map onderdeel van je $env:PSModulePath.
* VersionedOutputDirectory
  * Bij voorkeur staat deze op _$true_ zodat bij het aanpassen van het versie nummer in de betreffende bestanden er voor zorgt dat er een aparte map met versienummer gemaakt wordt waarin de module wordt opgeslagen. Op die manier heb je diverse versies beschikbaar en kun je altijd terug naar een vorige (werkende) versie van je module.

### Voorbeeld als module ModuleBuilder in je $env:PSModulePath staat

```PowerShell
Import-Module ModuleBuilder
```

### Voorbeeld als module ModuleBuilder NIET in je $env:PSModulePath staat

```PowerShell
Import-Module <FullPathTo>ModuleBuilder<.psm1>
```

### Bouw de module

De _-Prefix_ en/of _-Suffix_ zijn optioneel en alleen nodig als je ze ook echt gemaakt hebt met een bepaald doel om 'voor' en 'na' je functies opgenomen te worden in het module script.

```PowerShell
Build-Module '<FullPathTo>\PSModule.ToDo\Source\' [-Prefix 'prefix.ps1'] [-Suffix 'suffix.ps1']
```

---

## Uiteindelijke module

De module wordt gemaakt en geplaatst in de map die beschreven is in de 'build.psd1'.

### Voorbeeld van de module structuur

Gebaseerd op de bovenstaande parameters _OutputDirectory_ en _VersionedOutputDirectory_ is op __$true__ gezet en je hebt 2 versies '1.0.0' and '1.1.0' gemaakt de uiteindelijke structuur zal op het volgende voorbeeld lijken.

Als de module in je pad _$env:PSModulePath_ staat zal het commando _'PS> Import-Module PSModule.ToDo'_ altijd de laatste versie, in dit geval _'1.1.0'_ laden.

```PowerShell
PS> Get-ChildItem -Recurse "$( $env:OneDrive )\Modules\CultuurhuysDeKroon"

    Directory: <FullPathTo>\Modules\CultuurhuysDeKroon

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
la---            1-1-2020    17:31                1.0.0
la---           10-2-2020    21:04                1.0.1

    Directory: <FullPathTo>\Modules\CultuurhuysDeKroon\1.0.0

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
la---            1-1-2020    17:31                en-US
la---            1-1-2020    17:31                nl-NL
la---           29-1-2020    19:03           7574 CultuurhuysDeKroon.psd1
la---           30-1-2020    22:41           6694 CultuurhuysDeKroon.psm1

    Directory: <FullPathTo>\Modules\CultuurhuysDeKroon\1.0.1

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
la---           10-2-2020    21:04                en-US
la---           10-2-2020    21:04                nl-NL
la---           10-2-2020    21:04                Tests
la---           10-2-2020    21:04           4683 CultuurhuysDeKroon.psd1
la---            3-3-2020    20:55          96021 CultuurhuysDeKroon.psm1
```

* Map: en-US
  * Bevat de US taal specifieke help.
* Map: xx-XX
  * Bevat de XX taal specifieke help.
* Map: Private
  * Alle scripts die nodig zijn voor de module zelf maar deze worden niet aan de gebruiker getoond.
* Map: Public
  * Alle scripts die worden getoond aan de gerbruikers.
* Map: Tests
  * Deze Map bevat PowerShell [Pester](https://github.com/pester/Pester) tests voor het controleren/testen van je module en/of scripts.
* Bestand: PSModule.ToDo.psd1
  * Het module manifest bestand.
* Bestand: PSModule.ToDo.psm1
  * Het uiteindelijke module bestand dat wordt gemaakt door _Build-Module_ en is een samenvoeging van:
  * Prefix.ps1
    * Zoals opgegeven tijdens het bouwen in het _Build-Module_ commando '-prefix prefix.ps1'.
    * e.g. Dit zet een aantal zaken voordat functions worden gedefinieerd.
  * Alle functies
    * Alle functies gedefinieerd in de _private_ en _public_ mappen.
  * Suffix.ps1
    * Zoals opgegeven tijdens het bouwen in het _Build-Module_ commando '-suffix suffix.ps1'.
    * e.g. Alles wat je alvast automatisch wilt laten uitvoeren als alle functies gedefinieerd zijn.

---

## Mappen en bestanden in de source repository

In the root van de repo:

* Bestand - CHANGELOG.md
  * Om alle wijzigingen in bij te houden. Zelf houd ik hier alle wijzigingen in bij van alle ps1, psd1, psm1 bestanden.
* Bestand- README.md (dit bestand)
  * Algemene uitleg van de module.

---

## Map - Sources

Deze map bevat de bestanden die nodig zijn voor het bouwen van de module.

* Bestand - Build.psd1
  * Configuratie bestand voor ['ModuleBuilder'](https://github.com/PoshCode/ModuleBuilder)
* Bestand - Prefix.ps1
  * Gebruikt door ModuleBuilder.
  * De inhoud van dit bestand zal in het begin van _CultuurhuysDeKroon.psm1_ komen.
  * Hier kun je een aantal zaken definieren die je verderop nodig hebt in het module bestand. Bijvoorbeeld het laden van een configuratie bestand.
* Bestand - Suffix.ps1
  * Gebruikt door ModuleBuilder.
  * De inhoud van dit bestand zal in het eind van _CultuurhuysDeKroon.psm1_ komen na alle functies.
  * Hier kun je bijvoorbeeld alvast zaken uit laten voeren of wat opschoonwerk doen.
* Bestand - PSModule-Base.psd1
  * Het manifest bestand voor de module.

---

## Markdown guides

Hier een aantal verwijzingen naar documentatie om dit soort 'MD' (Markdown) bestanden te maken.

* [Markdown guide](https://www.markdownguide.org/basic-syntax/)
* [John Gruder - markdown creator](https://daringfireball.net/projects/markdown/)
* [GitHub mastering markdown](https://guides.github.com/features/mastering-markdown/)
* [Pandoc - PowerShell module to convert 'Markdown' to ...](http://pandoc.org/)

---

## Getest op

Operation System | PowerShell | PowerShell Core
-------------------------------------------------- | ---------- | ----------
Microsoft Windows 10 Home | 5.1 | 7.0, 7.1


Source/CHANGELOG.md
# Changelog voor Module CultuurhuysDeKroon

Het streven is om hierin alle wijzigingen bij te houden, ook voor alle scripts.

---

## ToDo

Waar nodig bestanden aanpassen zodat ze (ook) werken met mijn versie van 'Get-Command -Module <modulenaam>'.

```PowerShell
PS>Get-CommandVersion -Modules CultuurhuysDeKroon

Source (version)      Name                       Version Error
----------------      ----                       ------- -----
CultuurhuysDeKroon (1.2.3) Convert-AsciiTextViceversa 0.9.2
CultuurhuysDeKroon (1.2.3) Get-MACAddressVendor       0.0.0   Has no (proper) parameter "Version".
...
CultuurhuysDeKroon (1.2.3) Test-ArrayVsHashtable      0.0.0   Has no (proper) parameter "Version".
```

---

## Historie van wijzigingen

### 0.3.0 ( 2022-06-25 )

* _Nieuw_
  * Convert-TicketmaticExcel - Converteer een Ticketmatic MS-Excel export en maak een kopie met kleuren.

### 0.2.0 ( 2022-06-25 )

* _Aangepast_
  * Kleine type fouten.
  * Kleine cosmetische aanpassingen in een aantal script.
  * Nog 2-3 'PSModule.ToDo' vervangen door 'CultuurhuysDeKroon'.

### 0.1.0 ( 2022-06-24 )

* _Nieuw_
  * De basis mappen (Source, nl-NL, Tests, ...) en bestanden (CHANGELOG.MD, Build.ps1, ...).

---

## Types van wijzigingen

Type | Omschrijving
----- | -----
_Added_ | Nieuw.
_Changed_ | Wijziging.
_Deprecated_ | Wanneer bestanden (binnenkort) verwijderd worden.
_Removed_ | Verwijdert.
_Fixed_ | Voor fouten (bugs) die zijn aangepast.
_Security_ | Aanpassingen in relatie tot beveiliging.


Source/Prefix.ps1
<#
.SYNOPSIS
Prefix part to a module.

.DESCRIPTION
Prefix part to a module.
Based on ModuleBuilder documentation.
Use this prefix part to do some initialization for the module.
Like : load config for this module.

.EXAMPLE
NA

.INPUTS
NA

.OUTPUTS
NA

.NOTES
Copyright 2018-<today>, Marcel Rijsbergen.
    
History:
Do not forget to create a short modification text in CHANGELOG.md.

190518 MR
- 1.0.3 Added VerbosePreference check.

190518 MR
- 1.0.2 Still struggling with -verbose.

#>

#Region 'Initialize.'

[ CmdletBinding( )]

param( )

$CurCommand   = $MyInvocation.MyCommand
$ScriptName   = [io.path]::GetFileNameWithoutExtension($CurCommand.Name)
$ScriptFolder = Split-Path $CurCommand.Path
#TODO Find a way to check language for local name of 'Documents'.
[string] $FolderRoot   = Join-Path -Path $env:OneDrive -ChildPath 'Documenten'
[string] $FolderConfig = Join-Path -Path $FolderRoot -ChildPath 'Config'
[bool]   $IsVerbose    = $false

#Region : 'Check if Verbose file is set.'
$VerboseFile = $ScriptFolder + '\' + $ScriptName + '.On'
if (Test-Path $VerboseFile) {
    $IsVerbose = $true
}
#EndRegion : 'Check if Verbose file is set.'

#Region 'Function - Required - Get-TimeStamp.'
if (Get-Command Get-TimeStamp -ErrorAction SilentlyContinue) {
    if ($IsVerbose) {
        Write-Output -InputObject "$( Get-TimeStamp ) $( $ScriptName ) INFO Function 'Get-TimeStamp' exists."
    }
} else {
    if ($IsVerbose) {
        Write-Output -InputObject '...' $ScriptName "INFO Function 'Get-TimeStamp' does not exist, create a quick one."
    }
    function Get-TimeStamp {
        <#

        .SYNOPSIS
        Return a time stamp to be used in e.g.: logfiles, 

        .DESCRIPTION
        Return a time stamp to be used in e.g.: logfiles, 

        .EXAMPLE
        PS> Get-TimeStamp
        20191004-184821.729

        .INPUTS
        NA

        .OUTPUTS
        NA

        .NOTES
        Author      : Marcel Rijsbergen.
        Copyright   : 2018-<today>, Marcel Rijsbergen.
        History     : Can not remember. Used it already pretty quick after i created functions.
        Ideas       :
        - Add parameters to be able to return different output.
        
        #>
        Return (Get-Date -Format "yyyyMMdd-HHMMSS")
    }
}
#EndRegion 'Function - Required - Get-TimeStamp.'

if ($IsVerbose) {Write-Output -InputObject "$(Get-TimeStamp) $($ScriptName) INFO Loading module."}
#EndRegion 'Initialize.'

#Region 'Test for config and read it.'
$Curdir = Get-Location
if ($IsVerbose) {Write-Output -InputObject "$(Get-TimeStamp) $ScriptName INFO Current folder : $($curdir)"}
$ConfigFile = $FolderConfig + '\Config-' + $ScriptName + '.psd1'
$ConfigName = $ScriptName -replace '-','' -replace '_]','' -replace ' ',''
if ($IsVerbose) {Write-Output -InputObject "$(Get-TimeStamp) $ScriptName INFO Try Config file : $($ConfigFile)"}
if ($IsVerbose) {Write-Output -InputObject "$(Get-TimeStamp) $ScriptName INFO Set Config variable : $($ConfigName)"}
if (Test-Path $ConfigFile ) {
    try {
        $CurConfig = "Config.$($ConfigName)"
        $SetVarHash = @{
            Name        = $CurConfig
            Value       = (Import-PowerShellDataFile -Path $ConfigFile)
            Description = "Configuration for module $($ScriptName)"
        }
        Set-Variable -Scope Global @SetVarHash
        if ($IsVerbose) {
            Write-Host -Object "$(Get-TimeStamp) $ScriptName INFO Config loaded into : $($CurConfig)" -ForegroundColor Green
        }
        try {
            $ThisConfig = (Get-Variable $CurConfig).Value
            if ($ThisConfig.Folders) {
                if ($IsVerbose) {
                    Write-Output -InputObject "$(Get-TimeStamp) $ScriptName INFO Config found for folders, process."
                    Write-Output -InputObject "$(Get-TimeStamp) $ScriptName INFO Test folders."
                }
                else {
                    write-Host "$(Get-TimeStamp) $ScriptName INFO Process folder" -NoNewline
                }
                foreach ($TestFolder in $ThisConfig.Folders) {
                    $ThisFolder = Join-Path -Path $FolderRoot -ChildPath $TestFolder
                    if ($IsVerbose) {
                        Write-Output -InputObject "$(Get-TimeStamp) $ScriptName INFO Process folder : $($ThisFolder)"
                    }
                    $ThisStatus = 'none'
                    if (Test-Path -Path $ThisFolder) {
                        if ($IsVerbose) {Write-Output -InputObject "$(Get-TimeStamp) $ScriptName INFO Folder exists."}
                        $ThisStatus = 'exists'
                        Set-Variable -Name "Folder$($TestFolder)" -Value "$($ThisFolder)" -Scope Global
                    }
                    else {
                        if ($IsVerbose) {Write-Output -InputObject "$(Get-TimeStamp) $ScriptName INFO Folder does NOT exists."}
                        try {
                            if ($IsVerbose) {Write-Output -InputObject "$(Get-TimeStamp) $ScriptName INFO Create folder."}
                            New-Item -Path $ThisFolder -ItemType Directory -ErrorAction Stop
                            if ($IsVerbose) {Write-Output -InputObject "$(Get-TimeStamp) $ScriptName INFO Created folder."}
                            Set-Variable -Name "Folder$($TestFolder)" -Value "$($ThisFolder)" -Scope Global
                            $ThisStatus = 'create - success'
                        }
                        catch {
                            if ($IsVerbose) {Write-Warning -Message "$(Get-TimeStamp) $ScriptName INFO Error creating folder."}
                            $ThisStatus = 'create - failed'
                        }
                    }
                    if (-NOT ($IsVerbose)) {Write-Output -InputObject " - $($TestFolder) ($($ThisStatus))" -NoNewline}
                }
                if (-NOT ($IsVerbose)) {Write-Output -InputObject ''}
            }
            else {
                Write-Warning -Message "$(Get-TimeStamp) $($ScriptName) WARNING No config found for 'Folders'."
            }
        }
        catch {
            #
        }
    } catch {
        if ($IsVerbose) {
            Write-Host -Object "$(Get-TimeStamp) $ScriptName WARNING Config load : Failed" -ForegroundColor Yellow
        }
    }
} else {
    if ($IsVerbose) {
        Write-Output -InputObject "$(Get-TimeStamp) $ScriptName INFO No config file :" $ConfigFile
    }
}
#EndRegion 'Test for config and read it.'

#TODO Pester test nog werkend maken dan een 1e versie klaar.


Source/Public/Convert-TicketmaticExcel.ps1
function Convert-TicketmaticExcel {
    <#
    .SYNOPSIS
    Use a Ticketmatic exported 'Evenementen' MS-Excel file, convert data and color it.

    .DESCRIPTION
    Use a Ticketmatic exported 'Evenementen' MS-Excel file, convert data and color it.

    In Ticketmatic with the proper role you can export a list of 'Evenementen' to a MS-Excel file.
    Not sure if this is the default output, but in our case we get the following headers:
        Naam            Name of event.
        Ondertitel      Subtitle.
                        This is normally not used, only if the event is for 'customers' BDGV or VONK,
                        the values 'Bibliotheek De Groene Venen' or 'Stichting Vonk' should be used.
                        This will trigger the proper setup and colors in the mails send to the customers.
        Weekdag         Day of week.
        Datum           Date of event.
        Beschikbaarheid x/y     x = still available tickets, y = maximum tickets that can be sold.
        Status          If the event is in 'Draft' or 'Published' (active and visible).
        Genre(s)        There should be at least 1 (one) genre.


    .NOTES
    Name    : Convert-Ticketmatic
    Author  : Marcel Rijsbergen
    Historie:

    220625 - 1.1.0 MR
    - Added comment based help.
    - Made it a function.
    - Added some default PowerShell default setup like params, version.
    - Generate colors based on genre.
    - Colors are more close to the defined ones by the graphic designer(s): Ine, Inge
    - #TODO Split the genre and check if 'Bibliotheek' or 'VONK' is supplied. If 'Lezing' make it 'Bibliotheek'.

    220625 - 1.0.0 MR
    - Delivered a 1st working version.
    - Use the original MS-Excel sheet, put it into a named sheet and create a new sheet with a copy.
    - Tested with 2 hardcoded colors.

    220624 - 0.1.0 MR
    - Started based on the link below and several other (microsoft) pages about color, module ImportExcel, ...

    
    .LINK
    https://bronowski.it/blog/2020/12/how-to-format-an-entire-excel-row-based-on-the-cell-values-with-powershell/
    
    .EXAMPLE
    Test-MyTestFunction -Verbose
    
    #>
    
    #region 'Initialization.'
    [cmdletbinding()]
    param(
        [Parameter(Mandatory = $false)]
        [switch] $Version = $false
    )

    $ScriptName = [io.path]::GetFileNameWithoutExtension($MyInvocation.MyCommand.Name)
    Write-Verbose -Message "$(Get-TimeStamp) $($ScriptName) Begin."

    Write-Verbose -Message "$(Get-TimeStamp) $($ScriptName) Test if parameter 'Version' is supplied."
    [Version] $ScriptVersion = '1.1.0'
    if ($Version) {
        Write-Verbose -Message "$(Get-TimeStamp) $($ScriptName) Version: $($ScriptVersion)"
        return $ScriptVersion
    }
    #endregion 'Initialization.'

    # Colors used by graphic designer CDK.
    # Used calculator: https://www.w3schools.com/colors/colors_cmyk.asp
    #
    # #ff21ab   cmyk(0%, 87%, 33%, 0%)
    # #ffa81a   cmyk(0%, 34%, 90%, 0%)
    # #000000   cmyk(0%, 40%, 0%, 100%) -> cmyk(0%, 0%, 0%, 100%)
    # #33cccc   cmyk(80%, 20%, 20%, 100%) -> cmyk(0%, 0%, 0%, 100%)
    # #582314   cmyk(57%, 83%, 90%, 20%) -> cmyk(0%, 60%, 77%, 66%)
    # #ff5421   cmyk(0%, 67%, 87%, 0%)                                
    # #5ad081   cmyk(59%, 5%, 41%, 14%) -> cmyk(57%, 0%, 38%, 18%)
    # #9b00b6   cmyk(31%, 100%, 19%, 12%) -> cmyk(15%, 100%, 0%, 29%)
    # #2b6bff   cmyk(83%, 58%, 0%, 0%)                                
    #
    # CurColors:
    #   - The first colors are the same order as the CurTags. Colors from year calendar.
    #   - The remaining colors can be used for other purposes.
    $CurColorsRGB = @(
        # Colors for CurTags
        '#33cccc',  # rgb  51, 204, 204     cmyk(80%, 20%, 20%, 100%) -> cmyk(0%, 0%, 0%, 100%)
        '#ffa81a',  # rgb 255, 168,  26     cmyk(0%, 34%, 90%, 0%)
        '#ff5421',  # rgb 255,  84,  33     cmyk(0%, 67%, 87%, 0%)
        '#ff21ab',  # rgb 255,  33, 171     cmyk(0%, 87%, 33%, 0%)
        '#ffff00',  # rgb 255, 255,   0     cmyk(0%, 0%, 100%, 0%) - defined myself, seems yellow(ish).
        '#5ad081',  # rgb  90, 208, 129     cmyk(59%, 5%, 41%, 14%) -> cmyk(57%, 0%, 38%, 18%)

        # Default color for unknown CurTags.
        '#2b6bff',  # rgb  43, 107, 255     cmyk(83%, 58%, 0%, 0%)

        # Remaining colors.
        '#582314',  # rgb  88,  35,  20     cmyk(57%, 83%, 90%, 20%) -> cmyk(0%, 60%, 77%, 66%)
        '#9b00b6',  # rgb 155,   0, 182     cmyk(31%, 100%, 19%, 12%) -> cmyk(15%, 100%, 0%, 29%)
        '#000000'   # rgb   0,   0,   0     cmyk(0%, 40%, 0%, 100%) -> cmyk(0%, 0%, 0%, 100%)
    )
    $CurColors = @(
        'DarkCyan',
        'Gray'
        'Red',
        'DarkYellow',
        'DarkRed',
        'DarkGray',
        'Blue'
    )
    $CurTags = (
        'Cabaret',
        'Educatie',
        'Kinderen',
        'Lezing',
        'Muziek',
        'Pubquiz',
        'Toneel'
    )

    # Set variables.
    $MyLocation = 'D:\Users\rijsb\Downloads\CDK'
    $TMOriginal = Join-Path -Path $MyLocation -ChildPath 'Ticketmatic-export_20220620.xlsx'
    $TMImport   = Import-Excel -Path $TMOriginal
    $TMNew      = Join-Path -Path $MyLocation -ChildPath 'Ticketmatic-CultuurhuysDeKroon.xlsx'

    # Define sheet names.
    $SheetOriginal = 'Origineel'
    $SheetNew      = 'Nieuw'

    # Remove Excel working file.
    #Set-Location -Path $MyLocation
    write-verbose "Remove excel file if exists."
    if (Test-Path $TMNew) {
        Try {
            Remove-Item $TMNew -ErrorAction SilentlyContinue
        } Catch {
            write-Warning -Message "Error removing Excel work file, make sure the sheet is closed."
            Return
        }
    }

    # Connect to Excel file.
    Write-Output -InputObject "Create Excel file: $($TMNew)"
    $excelPackage = Open-ExcelPackage -Path $TMNew -KillExcel

    # Save original data to sheet 'Original' and close Excel file.
    Write-Output -InputObject "Save original data to sheet: $($SheetOriginal)"
    $TMImport | Export-Excel -Path $TMNew -WorksheetName $SheetOriginal -AutoFilter -AutoSize -KillExcel

    # Connect to Excel file.
    write-verbose "Connect to Excel file: $($TMNew)"
    $excelPackage = Open-ExcelPackage -Path $TMNew -KillExcel
    #write-verbose "Connect to sheet: $($SheetOriginal)"
    #$excelOriginal = $excelPackage.Workbook.Worksheets["$($SheetOriginal)"]

    # Close and show.
    #Close-ExcelPackage -ExcelPackage $excelPackage -Show

    # Save modified data to sheet 'New'.
    Write-Output -InputObject "Save original data to sheet: $($SheetNew)"
    $TMImport | Export-Excel -Path $TMNew -WorksheetName $SheetNew -AutoFilter -AutoSize -KillExcel

    # Connect to Excel file.
    write-verbose "Connect to Excel file: $($TMNew)"
    $excelPackage = Open-ExcelPackage -Path $TMNew -KillExcel
    write-verbose "Connect to sheet: $($SheetNew)"
    $excelNew = $excelPackage.Workbook.Worksheets["$($SheetNew)"]

    #$xx = $TMImport | Export-Excel -Path $TMNew -PassThru -ClearSheet
    #$ws        = $xx.Workbook.WorkSheets['Ticketmatic']
    #$totalRows = $ws.Dimension.Rows
    #$totalCols = $ws.Dimension.Cols
    #write-host -Object "Rows $($totalRows) Cols $($totalCols)"

    #$ws        = $excelNew.Workbook.WorkSheets["$($SheetNew)"]
    #$totalRows = $ws.Dimension.Rows
    #$totalCols = $ws.Dimension.Columns
    
    $totalRows = $excelNew.Dimension.Rows
    $totalCols = $excelNew.Dimension.Columns
    Write-Output -InputObject "Cols $($totalCols) Rows $($totalRows)"

    #works:     
    #$MyTag = "Muziek"
    #Add-ConditionalFormatting -Worksheet $excelNew -Address A2:g43 -RuleType Expression -ConditionValue '=$G2="Cabaret"' -BackgroundColor Blue
    #Add-ConditionalFormatting -Worksheet $excelNew -Address A2:g43 -RuleType Expression -ConditionValue ('=$G2="' + $MyTag + '"') -BackgroundColor DarkCyan
    
    foreach ($CurTag in $CurTags) {
        $TagIndex    = $CurTags.IndexOf("$($CurTag)")
        $TagColor    = $CurColors[$TagIndex]
        $TagColorRGB = $CurColorsRGB[$TagIndex] -replace '#', ''
        [int] $ColA = 8 # Transparency 255 is none.
        [int] $ColB = [Convert]::ToInt64($TagColorRGB.Substring(4, 2), 16)
        [int] $ColG = [Convert]::ToInt64($TagColorRGB.Substring(2, 2), 16)
        [int] $ColR = [Convert]::ToInt64($TagColorRGB.Substring(0, 2), 16)
        $NewBGColor = [System.Drawing.Color]::FromArgb($ColA, $ColB, $ColG, $ColR)
        Write-Output -InputObject "Genre: $($CurTag.PadRight(8)) - Color $($TagColor.PadRight(20)) - $($TagColorRGB) - $($ColA):$($ColR):$($ColG):$($ColB)"
        #Add-ConditionalFormatting -Worksheet $excelNew -Address A2:g43 -RuleType Expression -ConditionValue ('=$G2="' + $CurTag + '"') -BackgroundColor $TagColor
        Add-ConditionalFormatting -Worksheet $excelNew -Address A2:g43 -RuleType Expression -ConditionValue ('=$G2="' + $CurTag + '"') -BackgroundColor $NewBGColor
    }

    #Write-Output -InputObject "Set color to: $($MyTag)"
    #Add-ConditionalFormatting -Worksheet $excelNew -Address A2:g43 -RuleType Expression -ConditionValue "=$G2=""$($MyTag)""" -BackgroundColor "#ff21ab"

    #Set-ExcelRange -Range $ws.Cells["B2:B$($totalRows)"] -BackgroundColor LightBlue
    #Set-ExcelRange -Range $ws.Cells["a4:a$($totalCols)"] -BackgroundColor green

    # save the changes and open the spreadsheet
    Write-Output -InputObject "Close Excel file and show end result."
    Close-ExcelPackage -ExcelPackage $excelPackage -Show
    #Export-Excel -ExcelPackage $xx -KillExcel -Show
}
